### ISSUE TITLE
Build fails (use of deleted function)

### ISSUE THREAD

**Issue by vflyson** [rwengine/openrw#643]
OS: `Linux localhost 4.9.0-1-amd64 #1 SMP Debian 4.9.6-3 (2017-01-28) x86_64 GNU/Linux`
Compiler: `gcc version 6.3.0 20170516 (Debian 6.3.0-18+deb9u1)`

The log:
```
[ 65%] Building CXX object rwengine/CMakeFiles/rwengine.dir/src/render/GameRenderer.cpp.o
In file included from /usr/include/x86_64-linux-gnu/c++/6/bits/c++allocator.h:33:0,
                 from /usr/include/c++/6/bits/allocator.h:46,
                 from /usr/include/c++/6/memory:63,
                 from /home/vflyson/Projects/openrw/rwengine/src/objects/VehicleObject.hpp:6,
                 from /home/vflyson/Projects/openrw/rwengine/src/objects/VehicleObject.cpp:1:
/usr/include/c++/6/ext/new_allocator.h: In instantiation of ‘void __gnu_cxx::new_allocator<_Tp>::construct(_Up*, _Args&& ...) [with _Up = std::pair<const std::__cxx11::basic_string<char>, VehicleObject::Part>; _Args = {const std::pair<const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, VehicleObject::Part>&}; _Tp = std::pair<const std::__cxx11::basic_string<char>, VehicleObject::Part>]’:
/usr/include/c++/6/bits/alloc_traits.h:475:4:   required from ‘static void std::allocator_traits<std::allocator<_Tp1> >::construct(std::allocator_traits<std::allocator<_Tp1> >::allocator_type&, _Up*, _Args&& ...) [with _Up = std::pair<const std::__cxx11::basic_string<char>, VehicleObject::Part>; _Args = {const std::pair<const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, VehicleObject::Part>&}; _Tp = std::pair<const std::__cxx11::basic_string<char>, VehicleObject::Part>; std::allocator_traits<std::allocator<_Tp1> >::allocator_type = std::allocator<std::pair<const std::__cxx11::basic_string<char>, VehicleObject::Part> >]’
/usr/include/c++/6/bits/hashtable_policy.h:1953:37:   required from ‘std::__detail::_Hashtable_alloc<_NodeAlloc>::__node_type* std::__detail::_Hashtable_alloc<_NodeAlloc>::_M_allocate_node(_Args&& ...) [with _Args = {const std::pair<const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, VehicleObject::Part>&}; _NodeAlloc = std::allocator<std::__detail::_Hash_node<std::pair<const std::__cxx11::basic_string<char>, VehicleObject::Part>, true> >; std::__detail::_Hashtable_alloc<_NodeAlloc>::__node_type = std::__detail::_Hash_node<std::pair<const std::__cxx11::basic_string<char>, VehicleObject::Part>, true>]’
/usr/include/c++/6/bits/hashtable_policy.h:180:58:   required from ‘std::__detail::_AllocNode<_NodeAlloc>::__node_type* std::__detail::_AllocNode<_NodeAlloc>::operator()(_Arg&&) const [with _Arg = const std::pair<const std::__cxx11::basic_string<char>, VehicleObject::Part>&; _NodeAlloc = std::allocator<std::__detail::_Hash_node<std::pair<const std::__cxx11::basic_string<char>, VehicleObject::Part>, true> >; std::__detail::_AllocNode<_NodeAlloc>::__node_type = std::__detail::_Hash_node<std::pair<const std::__cxx11::basic_string<char>, VehicleObject::Part>, true>]’
/usr/include/c++/6/bits/hashtable.h:1690:18:   required from ‘std::pair<typename std::__detail::_Hashtable_base<_Key, _Value, _ExtractKey, _Equal, _H1, _H2, _Hash, _Traits>::iterator, bool> std::_Hashtable<_Key, _Value, _Alloc, _ExtractKey, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits>::_M_insert(_Arg&&, const _NodeGenerator&, std::true_type) [with _Arg = const std::pair<const std::__cxx11::basic_string<char>, VehicleObject::Part>&; _NodeGenerator = std::__detail::_AllocNode<std::allocator<std::__detail::_Hash_node<std::pair<const std::__cxx11::basic_string<char>, VehicleObject::Part>, true> > >; _Key = std::__cxx11::basic_string<char>; _Value = std::pair<const std::__cxx11::basic_string<char>, VehicleObject::Part>; _Alloc = std::allocator<std::pair<const std::__cxx11::basic_string<char>, VehicleObject::Part> >; _ExtractKey = std::__detail::_Select1st; _Equal = std::equal_to<std::__cxx11::basic_string<char> >; _H1 = std::hash<std::__cxx11::basic_string<char> >; _H2 = std::__detail::_Mod_range_hashing; _Hash = std::__detail::_Default_ranged_hash; _RehashPolicy = std::__detail::_Prime_rehash_policy; _Traits = std::__detail::_Hashtable_traits<true, false, true>; typename std::__detail::_Hashtable_base<_Key, _Value, _ExtractKey, _Equal, _H1, _H2, _Hash, _Traits>::iterator = std::__detail::_Node_iterator<std::pair<const std::__cxx11::basic_string<char>, VehicleObject::Part>, false, true>; std::true_type = std::integral_constant<bool, true>]’
/usr/include/c++/6/bits/hashtable_policy.h:713:55:   required from ‘std::__detail::_Insert_base<_Key, _Value, _Alloc, _ExtractKey, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits>::__ireturn_type std::__detail::_Insert_base<_Key, _Value, _Alloc, _ExtractKey, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits>::insert(const value_type&) [with _Key = std::__cxx11::basic_string<char>; _Value = std::pair<const std::__cxx11::basic_string<char>, VehicleObject::Part>; _Alloc = std::allocator<std::pair<const std::__cxx11::basic_string<char>, VehicleObject::Part> >; _ExtractKey = std::__detail::_Select1st; _Equal = std::equal_to<std::__cxx11::basic_string<char> >; _H1 = std::hash<std::__cxx11::basic_string<char> >; _H2 = std::__detail::_Mod_range_hashing; _Hash = std::__detail::_Default_ranged_hash; _RehashPolicy = std::__detail::_Prime_rehash_policy; _Traits = std::__detail::_Hashtable_traits<true, false, true>; std::__detail::_Insert_base<_Key, _Value, _Alloc, _ExtractKey, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits>::__ireturn_type = std::pair<std::__detail::_Node_iterator<std::pair<const std::__cxx11::basic_string<char>, VehicleObject::Part>, false, true>, bool>; std::__detail::_Insert_base<_Key, _Value, _Alloc, _ExtractKey, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits>::value_type = std::pair<const std::__cxx11::basic_string<char>, VehicleObject::Part>]’
/usr/include/c++/6/bits/unordered_map.h:550:31:   required from ‘std::pair<typename std::_Hashtable<_Key, std::pair<const _Key, _Tp>, _Alloc, std::__detail::_Select1st, _Pred, _Hash, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<std::__not_<std::__and_<std::__is_fast_hash<_Hash>, std::__detail::__is_noexcept_hash<_Key, _Hash> > >::value, false, true> >::iterator, bool> std::unordered_map<_Key, _Tp, _Hash, _Pred, _Alloc>::insert(const value_type&) [with _Key = std::__cxx11::basic_string<char>; _Tp = VehicleObject::Part; _Hash = std::hash<std::__cxx11::basic_string<char> >; _Pred = std::equal_to<std::__cxx11::basic_string<char> >; _Alloc = std::allocator<std::pair<const std::__cxx11::basic_string<char>, VehicleObject::Part> >; typename std::_Hashtable<_Key, std::pair<const _Key, _Tp>, _Alloc, std::__detail::_Select1st, _Pred, _Hash, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<std::__not_<std::__and_<std::__is_fast_hash<_Hash>, std::__detail::__is_noexcept_hash<_Key, _Hash> > >::value, false, true> >::iterator = std::__detail::_Node_iterator<std::pair<const std::__cxx11::basic_string<char>, VehicleObject::Part>, false, true>; std::unordered_map<_Key, _Tp, _Hash, _Pred, _Alloc>::value_type = std::pair<const std::__cxx11::basic_string<char>, VehicleObject::Part>]’
/home/vflyson/Projects/openrw/rwengine/src/objects/VehicleObject.cpp:886:42:   required from here
/usr/include/c++/6/ext/new_allocator.h:120:4: error: use of deleted function ‘constexpr std::pair<_T1, _T2>::pair(const std::pair<_T1, _T2>&) [with _T1 = const std::__cxx11::basic_string<char>; _T2 = VehicleObject::Part]’
  { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/6/bits/stl_algobase.h:64:0,
                 from /usr/include/c++/6/memory:62,
                 from /home/vflyson/Projects/openrw/rwengine/src/objects/VehicleObject.hpp:6,
                 from /home/vflyson/Projects/openrw/rwengine/src/objects/VehicleObject.cpp:1:
/usr/include/c++/6/bits/stl_pair.h:288:17: note: ‘constexpr std::pair<_T1, _T2>::pair(const std::pair<_T1, _T2>&) [with _T1 = const std::__cxx11::basic_string<char>; _T2 = VehicleObject::Part]’ is implicitly deleted because the default definition would be ill-formed:
       constexpr pair(const pair&) = default;
                 ^~~~
/usr/include/c++/6/bits/stl_pair.h:288:17: error: use of deleted function ‘VehicleObject::Part::Part(const VehicleObject::Part&)’
In file included from /home/vflyson/Projects/openrw/rwengine/src/objects/VehicleObject.cpp:1:0:
/home/vflyson/Projects/openrw/rwengine/src/objects/VehicleObject.hpp:64:12: note: ‘VehicleObject::Part::Part(const VehicleObject::Part&)’ is implicitly declared as deleted because ‘VehicleObject::Part’ declares a move constructor or move assignment operator
     struct Part {
            ^~~~
[ 66%] Building CXX object rwengine/CMakeFiles/rwengine.dir/src/render/GameShaders.cpp.o
[ 67%] Building CXX object rwengine/CMakeFiles/rwengine.dir/src/render/MapRenderer.cpp.o
rwengine/CMakeFiles/rwengine.dir/build.make:608: recipe for target 'rwengine/CMakeFiles/rwengine.dir/src/objects/VehicleObject.cpp.o' failed
make[2]: *** [rwengine/CMakeFiles/rwengine.dir/src/objects/VehicleObject.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
CMakeFiles/Makefile2:163: recipe for target 'rwengine/CMakeFiles/rwengine.dir/all' failed
make[1]: *** [rwengine/CMakeFiles/rwengine.dir/all] Error 2
Makefile:151: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by ShFil119** [comment]
Can you check clang or newer version of clang?

**Comment by madebr** [comment]
Just add `Part(Part& part) = default;` or remove https://github.com/rwengine/openrw/blob/0ae2a836ca99eaa23a6adcf4e85e4ae92b2de17a/rwengine/src/objects/VehicleObject.hpp#L83-L86 ?

**Comment by ShFil119** [comment]
But there shouldn't be need for that. Looks like compiler bug.

**Comment by madebr** [comment]
Upgrade to gcc 6.4 (or 7.3 or 8.2)
See https://gcc.gnu.org/

**Comment by vflyson** [comment]
It's the current on the stable Debian, I'd think a lot of people may get this error even if the gcc upgrade resolves it for me.

**Comment by Lihis** [comment]
On current stable (stretch) the building fails already to the CMake version check as minimum version requirement was bumped some time ago. Version bump consequences for Debian were acknowledged at least on IRC.

**Comment by ShFil119** [comment]
If we will switch to c++17, it would require even newer compiler.

You can avoid problem using backports or switching to newer release of distro.

**Comment by madebr** [comment]
Maybe add debian to the ci matrix?

**Comment by ShFil119** [comment]
@madebr I'd say ubuntu is like debian experimental.

**Comment by dracc** [comment]
@shfil More like Debian Testing. :)

**Comment by Lihis** [comment]
> switching to newer release of distro.

Stretch is the newest stable release :)

**Comment by vflyson** [comment]
I used backports for CMake before, apparently I ended up in a situations where most folks don't end up in, I call it edge case and respectfully close the issue.

**Comment by vflyson** [comment]
Oh, and thank you everyone Ü

**Comment by ShFil119** [comment]
Just in case, I'm building gcc 6.4 to verify if there's need for additional patches.

**Comment by ShFil119** [comment]
Happens also on 6.4. 
I've submitted bug on bugzilla:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=87658

**Comment by ShFil119** [comment]
Should be fixed on master.

**Comment by vflyson** [comment]
This got fixed by #648, thank  you @ShFil119
